drv/stm32g0-sys/src/main.rs

```diff
@@ -257,24 +158,7 @@ impl idl::InOrderSysImpl for ServerImpl<'_> {
         pins: u16,
         packed_attributes: u16,
     ) -> Result<(), RequestError<GpioError>> {
-        match port {
-            Port::A => {
-                let gpio = unsafe { &*device::GPIOA::ptr() };
-                do_configure!(gpio, pins, packed_attributes);
-            }
-            _ => {
-                let gpio = match port {
-                    Port::B => unsafe { &*device::GPIOB::ptr() },
-                    Port::C => unsafe { &*device::GPIOC::ptr() },
-                    Port::D => unsafe { &*device::GPIOD::ptr() },
-                    #[cfg(feature = "g0b1")]
-                    Port::E => unsafe { &*device::GPIOE::ptr() },
-                    Port::F => unsafe { &*device::GPIOF::ptr() },
-                    _ => unreachable!(),
-                };
-                do_configure!(gpio, pins, packed_attributes);
-            }
-        }
+        unsafe { get_gpio_regs(port) }.configure(pins, packed_attributes);
```

Comment:
siiiiiiiiigh

So.

The family of APIs that grew up around svd2rust decided, early on, to try to manage peripheral ownership using little "tokens" that can't be duplicated. You do `Peripherals::take()` at boot and it gives you a struct of tokens, and you can access peripherals through those tokens. This approach is useful in some contexts.

We don't use it, though, because it assumes that all code is in the same address space and is simply unsound in unprivileged isolated mode.

This means we're sort of on the gray areas of the svd2rust-generated APIs. In particular, they decided that any access to peripherals _outside_ of this token scheme would be unsafe. This is an abuse of `unsafe` to mean something other than `unsafe` -- technically there are peripherals one could poke that could violate memory safety, like anything capable of remapping memory or a DMA controller, but they are exceptions and the GPIO peripheral is certainly not among them.

So, a side effect of their decision is that there's no static `GPIOA` item defined, to prevent "ambient access" to the GPIOA peripheral bypassing the token scheme. Instead, all they'll do is give us a raw pointer from the `GPIOA::ptr()` call, which is the only call we can make without holding one of their tokens. We have to indirect it to get at the actual `RegisterBlock`, which is unsafe-in-the-Rust-sense.

Now, in this case, why did I also make `get_gpio_regs` unsafe, rather than hiding the `unsafe` within it? I could go either way on that.

Reply:
Got it, that all makes sense.  I don't feel strongly about this, so matching `svd2rust`'s decisions (however misguided) seems reasonable.

Reply:
(This is related to the reason why accessing raw `bits` is `unsafe` on _some_ registers and not others. Basically the default is to make any bit manipulation `unsafe`; someone has to go through and edit a giant YAML file to mark operations as ok to do from safe code. People have not done that for our processors. It gets really fun when they've done it for _some_ of our processors and not others, which gets us "unnecessary unsafe" warnings when compiling for some targets. This is part of why I'm concerned that our use case is diverging too far from what the PAC authors intend.)